### PR TITLE
[Feature] enable exporting to onnx for PointRend

### PR DESCRIPTION
**Pull request by DmitriySidnev**
This PR with [PR#953](https://github.com/open-mmlab/mmcv/pull/953) in mmcv enables exporting to ONNX for `PointRend` based models. Affected models:
* [point_rend/point_rend_r50_caffe_fpn_mstrain_1x_coco](https://github.com/open-mmlab/mmdetection/blob/master/configs/point_rend/point_rend_r50_caffe_fpn_mstrain_1x_coco.py)
* [point_rend/point_rend_r50_caffe_fpn_mstrain_3x_coco](https://github.com/open-mmlab/mmdetection/blob/master/configs/point_rend/point_rend_r50_caffe_fpn_mstrain_3x_coco.py)

### CODE REVIEW

**Comment by RunningLeon** [comment]
@DmitriySidnev Thanks for the contribution. Please list the models you have verified in the PR description.

**Comment by QingChuanWS** [comment]
Please provide the version of pytorch you used when testing these models.

**Comment by DmitriySidnev** [comment]
@QingChuanWS, hi! I use pytorch 1.8.1.

**Comment by QingChuanWS** [comment]
Hi, @DmitriySidnev. Have you verified whether the visualized result of the modified point_rend under pytorch is correct?

File: mmdet/models/roi_heads/mask_heads/mask_point_head.py [hunk 1]
```diff
@@ -285,16 +285,19 @@ def get_roi_rel_points_test(self, mask_pred, pred_label, cfg):
...
         uncertainty_map = uncertainty_map.view(num_rois,
                                                mask_height * mask_width)
         num_points = min(mask_height * mask_width, num_points)
         point_indices = uncertainty_map.topk(num_points, dim=1)[1]
-        point_coords = uncertainty_map.new_zeros(num_rois, num_points, 2)
-        point_coords[:, :, 0] = w_step / 2.0 + (point_indices %
-                                                mask_width).float() * w_step
-        point_coords[:, :, 1] = h_step / 2.0 + (point_indices //
-                                                mask_width).float() * h_step
+        a = w_step / 2.0 + (point_indices.long() % mask_width).float() * w_step
```
Comment (RunningLeon) [inline mmdet/models/roi_heads/mask_heads/mask_point_head.py:299]: Could use meaningful names for the variable `a` and `b`.

Comment (DmitriySidnev) [reply mmdet/models/roi_heads/mask_heads/mask_point_head.py]: Renamed.

**Comment by DmitriySidnev** [comment]
@QingChuanWS, I have verified both metrics and visualized results.

**Comment by RunningLeon** [comment]
@DmitriySidnev Could you merge with master and test if it is OK for Point Rend model exporting to ONNX with batch and dynamic shape if possible?

**Comment by DmitriySidnev** [comment]
@RunningLeon, I have checked exporting to ONNX with dynamic shape and batch. After fixing a bug in `mmcv/ops/point_sample/bilinear_grid_sample` function it works. But there is the line `if torch.onnx.is_in_onnx_export() and num_imgs == 1:` in my code that brakes logic with batch during exporting. It can be simply fixed via exporting with input tensor with first dimension (batch_size) = 2. The result onnx graph is valid.

**Comment by RunningLeon** [comment]
@DmitriySidnev Hi, please allow me to push for this PR.
```
ERROR: Permission to DmitriySidnev/mmdetection.git denied to RunningLeon.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

**Comment by DmitriySidnev** [comment]
@RunningLeon, hi! Access permissions granted.

**Comment by RangiLyu** [comment]
Hi~
I tried to test `point_rend_r50_caffe_fpn_mstrain_1x_coco` onnx model with onnxruntime, but run out of memory.

error message:

```

Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_2702' Status Message: /onnxruntime_src/onnxruntime/core/framework/bfc_arena.cc:305 void* onnxruntime::BFCArena::AllocateRawInternal(size_t, bool) Failed to allocate memory for requested buffer of size 1605632000

Traceback (most recent call last):
  File "tools/deployment/test.py", line 132, in <module>
    main()
  File "tools/deployment/test.py", line 111, in main
    args.show_score_thr)
  File "/home/rangilyu/Projects/mmdetection/mmdet/apis/test.py", line 27, in single_gpu_test
    result = model(return_loss=False, rescale=True, **data)
  File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/rangilyu/Projects/mmcv/mmcv/parallel/data_parallel.py", line 42, in forward
    return super().forward(*inputs, **kwargs)
  File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 165, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/rangilyu/Projects/mmcv/mmcv/runner/fp16_utils.py", line 95, in new_func
    return old_func(*args, **kwargs)
  File "/home/rangilyu/Projects/mmdetection/mmdet/models/detectors/base.py", line 169, in forward
    return self.forward_test(img, img_metas, **kwargs)
  File "/home/rangilyu/Projects/mmdetection/mmdet/core/export/model_wrappers.py", line 74, in forward_test
    self.sess.run_with_iobinding(self.io_binding)
  File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 229, in run_with_iobinding
    self._sess.run_with_iobinding(iobinding._iobinding, run_options)
RuntimeError: Error in execution: Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_2702' Status Message: /onnxruntime_src/onnxruntime/core/framework/bfc_arena.cc:305 void* onnxruntime::BFCArena::AllocateRawInternal(size_t, bool) Failed to allocate memory for requested buffer of size 1605632000

```
'ScatterElements_2702' failed to allocate memory for requested buffer of size 1605632000.

Why does this node `ScatterElements_2702` need so much memory?

**Comment by RangiLyu** [comment]
> Hi~
> I tried to test `point_rend_r50_caffe_fpn_mstrain_1x_coco` onnx model with onnxruntime, but run out of memory.
> 
> error message:
> 
> ```
> 
> Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_2702' Status Message: /onnxruntime_src/onnxruntime/core/framework/bfc_arena.cc:305 void* onnxruntime::BFCArena::AllocateRawInternal(size_t, bool) Failed to allocate memory for requested buffer of size 1605632000
> 
> Traceback (most recent call last):
>   File "tools/deployment/test.py", line 132, in <module>
>     main()
>   File "tools/deployment/test.py", line 111, in main
>     args.show_score_thr)
>   File "/home/rangilyu/Projects/mmdetection/mmdet/apis/test.py", line 27, in single_gpu_test
>     result = model(return_loss=False, rescale=True, **data)
>   File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
>     result = self.forward(*input, **kwargs)
>   File "/home/rangilyu/Projects/mmcv/mmcv/parallel/data_parallel.py", line 42, in forward
>     return super().forward(*inputs, **kwargs)
>   File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/torch/nn/parallel/data_parallel.py", line 165, in forward
>     return self.module(*inputs[0], **kwargs[0])
>   File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
>     result = self.forward(*input, **kwargs)
>   File "/home/rangilyu/Projects/mmcv/mmcv/runner/fp16_utils.py", line 95, in new_func
>     return old_func(*args, **kwargs)
>   File "/home/rangilyu/Projects/mmdetection/mmdet/models/detectors/base.py", line 169, in forward
>     return self.forward_test(img, img_metas, **kwargs)
>   File "/home/rangilyu/Projects/mmdetection/mmdet/core/export/model_wrappers.py", line 74, in forward_test
>     self.sess.run_with_iobinding(self.io_binding)
>   File "/home/rangilyu/anaconda3/envs/mmonnx/lib/python3.7/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 229, in run_with_iobinding
>     self._sess.run_with_iobinding(iobinding._iobinding, run_options)
> RuntimeError: Error in execution: Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_2702' Status Message: /onnxruntime_src/onnxruntime/core/framework/bfc_arena.cc:305 void* onnxruntime::BFCArena::AllocateRawInternal(size_t, bool) Failed to allocate memory for requested buffer of size 1605632000
> ```
> 
> 'ScatterElements_2702' failed to allocate memory for requested buffer of size 1605632000.
> 
> Why does this node `ScatterElements_2702` need so much memory?

Find a similar issue in onnxruntime(https://github.com/microsoft/onnxruntime/issues/7612). Switched to a lower version of onnxruntime and solved this problem.

**Comment by RangiLyu** [comment]
I have tested the performance of the ONNX model with onnxruntime. Here is the result:

|    Model     |  Type   | box mAP | box AP50| box AP75| mask mAP | mask AP50 | mask AP75 |
:--------------: | :------: | :------------: | :----: | :-----: | :------: | :--------: | :--------: |
point_rend_r50_caffe_fpn_mstrain_1x_coco| PyTorch | 38.4 | 59.0 | 41.8 | 36.3 | 56.9 | 38.7 |
point_rend_r50_caffe_fpn_mstrain_1x_coco| ONNX | 37.9 | 58.4 | 41.5| 34.9 | 55.9 | 36.8 |

The performance degradation of mask is large.

**Comment by DmitriySidnev** [comment]
@RangiLyu, could you please share scripts and commands which you use for export and test the model?

**Comment by RangiLyu** [comment]
> @RangiLyu, could you please share scripts and commands which you use for export and test the model?

```
python tools/deployment/pytorch2onnx.py configs/point_rend/point_rend_r50_caffe_fpn_mstrain_1x_coco.py checkpoints/point_rend_r50_caffe_fpn_mstrain_1x_coco-1bcb5fb4.pth --output-file  checkpoints/point_rend_r50_caffe_fpn_mstrain_1x_coco-1bcb5fb4.onnx --dynamic-export
```
and

```
python tools/deployment/test.py configs/point_rend/point_rend_r50_caffe_fpn_mstrain_1x_coco.py checkpoints/point_rend_r50_caffe_fpn_mstrain_1x_coco-1bcb5fb4.onnx --out work_dirs/point_rend_onnx/result.pkl --eval bbox segm
```

**Comment by DmitriySidnev** [comment]
@RangiLyu, I could not find the reason of metrics degradation. Only one thing that in my view may be related to the problem is a constant spatial resolution of output masks (by default it is 800x1216) in onnx graph. But I am not sure about it. On the other hand significant drop of segmentation metrics probably to be due to lower detection quality.

**Comment by RunningLeon** [comment]
@DmitriySidnev @RangiLyu Could refer to [here](https://github.com/open-mmlab/mmdetection/blob/master/docs/tutorials/pytorch2onnx.md#results-and-models) for possible reasons.

> Mask AP of Mask R-CNN drops by 1% for ONNXRuntime. The main reason is that the predicted masks are directly interpolated to original image in PyTorch, while they are at first interpolated to the preprocessed input image of the model and then to original image in ONNXRuntime.

**Comment by RunningLeon** [comment]
@DmitriySidnev , @RangiLyu https://github.com/open-mmlab/mmdetection/pull/5197 may be a possible reason.

**Comment by AronLin** [comment]
I merged this PR with the latest master branch, and tested the performance of the ONNX model with onnxruntime. 

Here is the result:

|    Model     |  Type   | box mAP | box AP50| box AP75| mask mAP | mask AP50 | mask AP75 |
:--------------: | :------: | :------------: | :----: | :-----: | :------: | :--------: | :--------: |
point_rend_r50_caffe_fpn_mstrain_1x_coco| PyTorch | 38.4 | 59.0 | 41.8 | 36.3 | 56.9 | 38.7 |
point_rend_r50_caffe_fpn_mstrain_1x_coco| ONNX | 38.4 | 59.0 | 41.8 | 35.2 | 56.5 | 37.1 |

It seems all right.

**Comment by RunningLeon** [comment]
Kindly ping@ZwwWayne

**Comment by RunningLeon** [comment]
@DmitriySidnev Hi, could you fix conflicts and refactor the onnx export according to #5205? Thanks a lot.

**Review by RangiLyu — APPROVED** [review]
LGTM

**Comment by ZwwWayne** [comment]
Can we put the complete ONNX export logic to an `onnx_export` function and separate them from the original inference code?
See example [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/roi_heads/standard_roi_head.py#L272).

**Comment by DmitriySidnev** [comment]
@ZwwWayne, I was worried about the functionality and now I really do not have enough free time to refactor the code. Can anyone else do this?

**Comment by ZwwWayne** [comment]
> @ZwwWayne, I was worried about the functionality and now I really do not have enough free time to refactor the code. Can anyone else do this?

OK, we will put effort on that. Thanks for the efforts.

**Comment by ZwwWayne** [comment]
Merged in #5440

**Pull request closed without merging**